File: src/lib/brand-guardian.ts
import type { BrandStyleJSON, FIBOPromptJSON, Violation, AutoFix } from '../types/database';

export interface ValidationResult {
  isValid: boolean;
  complianceScore: number;
  violations: Violation[];
  autoFixesAvailable: number;
}

export interface ValidationWithFixes extends ValidationResult {
  originalPrompt: FIBOPromptJSON;
  fixedPrompt: FIBOPromptJSON;
  appliedFixes: AutoFix[];
}

function hexToRgb(hex: string): { r: number; g: number; b: number } | null {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result ? {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16),
  } : null;
}

function colorDistance(hex1: string, hex2: string): number {
  const rgb1 = hexToRgb(hex1);
  const rgb2 = hexToRgb(hex2);
  if (!rgb1 || !rgb2) return Infinity;
  return Math.sqrt(
    Math.pow(rgb1.r - rgb2.r, 2) +
    Math.pow(rgb1.g - rgb2.g, 2) +
    Math.pow(rgb1.b - rgb2.b, 2)
  );
}

function findClosestColor(hex: string, palette: { hex: string; name: string }[]): { hex: string; name: string; distance: number } | null {
  if (palette.length === 0) return null;
  let closest = palette[0];
  let minDistance = colorDistance(hex, closest.hex);

  for (const color of palette) {
    const dist = colorDistance(hex, color.hex);
    if (dist < minDistance) {
      minDistance = dist;
      closest = color;
    }
  }

  return { ...closest, distance: minDistance };
}

function extractHexColors(colorScheme: string | undefined): string[] {
  if (!colorScheme || typeof colorScheme !== 'string') {
    return [];
  }
  const hexPattern = /#[0-9A-Fa-f]{6}/g;
  return colorScheme.match(hexPattern) || [];
}

function extractFocalLength(focalLength: string | undefined): number | null {
  if (!focalLength || typeof focalLength !== 'string') {
    return null;
  }
  const match = focalLength.match(/(\d+)\s*mm/i);
  return match ? parseInt(match[1], 10) : null;
}

function extractCameraAngle(cameraAngle: string | undefined): number | null {
  if (!cameraAngle || typeof cameraAngle !== 'string') {
    return null;
  }
  const match = cameraAngle.match(/(\d+)/);
  return match ? parseInt(match[1], 10) : null;
}

export function validateFIBOPrompt(
  prompt: FIBOPromptJSON,
  brandStyle: BrandStyleJSON
): ValidationResult {
  const violations: Violation[] = [];

  // Safely extract color_scheme (might be undefined or different type)
  const hexColors = extractHexColors(prompt.color_scheme as any);
  for (const hex of hexColors) {
    const isApproved = brandStyle.colorPalette.some(
      c => colorDistance(c.hex, hex) < 30
    );

    if (!isApproved) {
      const closest = findClosestColor(hex, brandStyle.colorPalette);
      // For trend-based collections, color differences are suggestions, not critical
      violations.push({
        id: crypto.randomUUID(),
        rule: 'Color differs from brand palette (OK for trends)',
        category: 'color',
        severity: 'suggestion', // Changed from 'critical' since trends may use different colors
        detected: hex,
        allowed: brandStyle.colorPalette.map(c => `${c.name} (${c.hex})`).join(', ') || 'No colors defined',
        message: `Trend color ${hex} differs from brand palette - this is acceptable for trend-based collections`,
        autoFixAvailable: closest !== null,
        fixedValue: closest?.hex,
      });
    }
  }

  const { cameraSettings } = brandStyle;
  const focalLength = extractFocalLength(prompt.focal_length as any);

  if (focalLength !== null) {
    const fovFromFocal = Math.round(2 * Math.atan(36 / (2 * focalLength)) * (180 / Math.PI));
    if (fovFromFocal < cameraSettings.fovMin || fovFromFocal > cameraSettings.fovMax) {
      const suggestedFocal = Math.round(36 / (2 * Math.tan((cameraSettings.fovDefault * Math.PI / 180) / 2)));
      violations.push({
        id: crypto.randomUUID(),
        rule: 'Focal length must produce FOV within allowed range',
        category: 'camera',
        severity: 'warning',
        detected: `${focalLength}mm (FOV ~${fovFromFocal}°)`,
        allowed: { min: cameraSettings.fovMin, max: cameraSettings.fovMax },
        message: `Focal length ${focalLength}mm produces FOV ~${fovFromFocal}° outside allowed range (${cameraSettings.fovMin}°-${cameraSettings.fovMax}°)`,
        autoFixAvailable: true,
        fixedValue: `${suggestedFocal}mm`,
      });
    }
  }

  const numericAngle = extractCameraAngle(prompt.camera_angle as any);
  if (numericAngle !== null) {
    if (numericAngle < cameraSettings.angleMin || numericAngle > cameraSettings.angleMax) {
      violations.push({
        id: crypto.randomUUID(),
        rule: 'Camera angle must be within allowed range',
        category: 'camera',
        severity: 'warning',
        detected: prompt.camera_angle,
        allowed: { min: cameraSettings.angleMin, max: cameraSettings.angleMax },
        message: `Camera angle "${prompt.camera_angle}" is outside allowed range (${cameraSettings.angleMin}°-${cameraSettings.angleMax}°)`,
        autoFixAvailable: true,
        fixedValue: `${cameraSettings.angleDefault}° eye level`,
      });
    }
  }

  const { lightingConfig } = brandStyle;
  // Handle lighting as either string or object
  let lightingLower = '';
  if (typeof prompt.lighting === 'string') {
    lightingLower = (prompt.lighting || '').toLowerCase();
  } else if (prompt.lighting && typeof prompt.lighting === 'object') {
    // If lighting is an object, combine its properties into a string
    lightingLower = JSON.stringify(prompt.lighting).toLowerCase();
  }

  if (lightingLower && lightingConfig.colorTemperature < 4500 && lightingLower.includes('cool')) {
    violations.push({
      id: crypto.randomUUID(),
      rule: 'Lighting color temperature should match brand style',
      category: 'lighting',
      severity: 'suggestion',
      detected: 'cool lighting',
      allowed: `warm lighting (${lightingConfig.colorTemperature}K)`,
      message: `Cool lighting detected but brand style specifies warm (${lightingConfig.colorTemperature}K)`,
      autoFixAvailable: true,
      fixedValue: 'warm, soft lighting with natural tones',
    });
  } else if (lightingLower && lightingConfig.colorTemperature > 5500 && lightingLower.includes('warm')) {
    violations.push({
      id: crypto.randomUUID(),
      rule: 'Lighting color temperature should match brand style',
      category: 'lighting',
      severity: 'suggestion',
      detected: 'warm lighting',
      allowed: `cool lighting (${lightingConfig.colorTemperature}K)`,
      message: `Warm lighting detected but brand style specifies cool (${lightingConfig.colorTemperature}K)`,
      autoFixAvailable: true,
      fixedValue: 'cool, neutral studio lighting',
    });
  }

  // Handle description as either string or object
  let descriptionLower = '';
  if (typeof prompt.description === 'string') {
    descriptionLower = (prompt.description || '').toLowerCase();
  } else if (prompt.description && typeof prompt.description === 'object') {
    descriptionLower = JSON.stringify(prompt.description).toLowerCase();
  }
  for (const negative of brandStyle.negativePrompts) {
    if (descriptionLower && descriptionLower.includes(negative.toLowerCase())) {
      violations.push({
        id: crypto.randomUUID(),
        rule: 'Description must not contain forbidden terms',
        category: 'prompt',
        severity: 'critical',
        detected: negative,
        allowed: 'Not allowed: ' + brandStyle.negativePrompts.join(', '),
        message: `Description contains forbidden term: "${negative}"`,
        autoFixAvailable: true,
        fixedValue: (prompt.description || '').replace(new RegExp(negative, 'gi'), '').trim(),
      });
    }
  }

  const currentNegatives = prompt.negative_prompt?.toLowerCase() || '';
  for (const negative of brandStyle.negativePrompts) {
    if (!currentNegatives.includes(negative.toLowerCase())) {
      violations.push({
        id: crypto.randomUUID(),
        rule: 'Negative prompt must include brand exclusions',
        category: 'prompt',
        severity: 'warning',
        detected: 'Missing: ' + negative,
        allowed: brandStyle.negativePrompts.join(', '),
        message: `Missing required negative prompt: "${negative}"`,
        autoFixAvailable: true,
        fixedValue: negative,
      });
    }
  }

  for (const obj of (prompt.objects || [])) {
    // Handle object description as either string or object
    let objDescLower = '';
    if (typeof obj.description === 'string') {
      objDescLower = (obj.description || '').toLowerCase();
    } else if (obj.description && typeof obj.description === 'object') {
      objDescLower = JSON.stringify(obj.description).toLowerCase();
    }
    for (const negative of brandStyle.negativePrompts) {
      if (objDescLower.includes(negative.toLowerCase())) {
        violations.push({
          id: crypto.randomUUID(),
          rule: 'Object description must not contain forbidden terms',
          category: 'prompt',
          severity: 'critical',
          detected: `"${negative}" in object "${obj.name}"`,
          allowed: 'Not allowed: ' + brandStyle.negativePrompts.join(', '),
          message: `Object "${obj.name}" description contains forbidden term: "${negative}"`,
          autoFixAvailable: true,
          fixedValue: (obj.description || '').replace(new RegExp(negative, 'gi'), '').trim(),
        });
      }
    }
  }

  const criticalCount = violations.filter(v => v.severity === 'critical').length;
  const warningCount = violations.filter(v => v.severity === 'warning').length;
  const suggestionCount = violations.filter(v => v.severity === 'suggestion').length;

  const maxScore = 100;
  const criticalPenalty = 25;
  const warningPenalty = 10;
  const suggestionPenalty = 3;

  const complianceScore = Math.max(0, Math.min(100,
    maxScore -
    (criticalCount * criticalPenalty) -
    (warningCount * warningPenalty) -
    (suggestionCount * suggestionPenalty)
  ));

  return {
    isValid: criticalCount === 0,
    complianceScore,
    violations,
    autoFixesAvailable: violations.filter(v => v.autoFixAvailable).length,
  };
}

export function applyAutoFixes(
  prompt: FIBOPromptJSON,
  brandStyle: BrandStyleJSON,
  validationResult: ValidationResult
): ValidationWithFixes {
  const fixedPrompt: FIBOPromptJSON = JSON.parse(JSON.stringify(prompt));
  const appliedFixes: AutoFix[] = [];

  for (const violation of validationResult.violations) {
    if (!violation.autoFixAvailable || violation.fixedValue === undefined) continue;

    const fix: AutoFix = {
      violationId: violation.id,
      field: '',
      originalValue: violation.detected,
      fixedValue: violation.fixedValue,
      appliedAt: new Date().toISOString(),
    };

    switch (violation.category) {
      case 'color':
        if (typeof violation.fixedValue === 'string' && typeof violation.detected === 'string' && fixedPrompt.color_scheme) {
          fix.field = 'color_scheme';
          fixedPrompt.color_scheme = fixedPrompt.color_scheme.replace(
            violation.detected,
            violation.fixedValue
          );
          appliedFixes.push(fix);
        }
        break;

      case 'camera':
        if (violation.rule.includes('Focal length') && typeof violation.fixedValue === 'string') {
          fix.field = 'focal_length';
          fixedPrompt.focal_length = violation.fixedValue;
          appliedFixes.push(fix);
        } else if (violation.rule.includes('angle') && typeof violation.fixedValue === 'string') {
          fix.field = 'camera_angle';
          fixedPrompt.camera_angle = violation.fixedValue;
          appliedFixes.push(fix);
        }
        break;

      case 'lighting':
        if (typeof violation.fixedValue === 'string') {
          fix.field = 'lighting';
          fixedPrompt.lighting = violation.fixedValue;
          appliedFixes.push(fix);
        }
        break;

      case 'prompt':
        if (violation.rule.includes('Description') && typeof violation.fixedValue === 'string') {
          fix.field = 'description';
          fixedPrompt.description = violation.fixedValue;
          appliedFixes.push(fix);
        } else if (violation.rule.includes('Negative prompt') && typeof violation.fixedValue === 'string') {
          fix.field = 'negative_prompt';
          const existing = fixedPrompt.negative_prompt || '';
          if (!existing.toLowerCase().includes(violation.fixedValue.toLowerCase())) {
            fixedPrompt.negative_prompt = existing
              ? `${existing}, ${violation.fixedValue}`
              : violation.fixedValue;
          }
          appliedFixes.push(fix);
        } else if (violation.rule.includes('Object description')) {
          for (let i = 0; i < fixedPrompt.objects.length; i++) {
            if (violation.message.includes(fixedPrompt.objects[i].name)) {
              fix.field = `objects[${i}].description`;
              fixedPrompt.objects[i].description = String(violation.fixedValue);
              appliedFixes.push(fix);
              break;
            }
          }
        }
        break;
    }
  }

  const newValidation = validateFIBOPrompt(fixedPrompt, brandStyle);

  return {
    ...newValidation,
    originalPrompt: prompt,
    fixedPrompt,
    appliedFixes,
  };
}

export function getComplianceBadge(score: number): { label: string; color: string; bgColor: string } {
  if (score >= 90) return { label: 'Excellent', color: 'text-emerald-400', bgColor: 'bg-emerald-500/20' };
  if (score >= 75) return { label: 'Good', color: 'text-green-400', bgColor: 'bg-green-500/20' };
  if (score >= 60) return { label: 'Fair', color: 'text-amber-400', bgColor: 'bg-amber-500/20' };
  if (score >= 40) return { label: 'Poor', color: 'text-orange-400', bgColor: 'bg-orange-500/20' };
  return { label: 'Critical', color: 'text-red-400', bgColor: 'bg-red-500/20' };
}

export function getSeverityIcon(severity: Violation['severity']): { icon: string; color: string } {
  switch (severity) {
    case 'critical': return { icon: 'XCircle', color: 'text-red-400' };
    case 'warning': return { icon: 'AlertTriangle', color: 'text-amber-400' };
    case 'suggestion': return { icon: 'Info', color: 'text-blue-400' };
  }
}

File: trendsync-backend/shared/brand_guardian.py
"""
Brand Guardian — Prompt & Visual Validation
Port of brand-guardian.ts to Python, plus new visual validation via Gemini 3 Flash.
"""

import math
import re
import uuid
from typing import Any, Dict, List, Optional, Tuple


# --------------------------------------------------------------------------
# Colour helpers
# --------------------------------------------------------------------------

def _hex_to_rgb(hex_str: str) -> Optional[Tuple[int, int, int]]:
    hex_str = hex_str.lstrip("#")
    if len(hex_str) != 6:
        return None
    try:
        return (
            int(hex_str[0:2], 16),
            int(hex_str[2:4], 16),
            int(hex_str[4:6], 16),
        )
    except ValueError:
        return None


def _color_distance(hex1: str, hex2: str) -> float:
    rgb1 = _hex_to_rgb(hex1)
    rgb2 = _hex_to_rgb(hex2)
    if rgb1 is None or rgb2 is None:
        return float("inf")
    return math.sqrt(sum((a - b) ** 2 for a, b in zip(rgb1, rgb2)))


def _find_closest_color(
    hex_val: str, palette: List[Dict[str, str]]
) -> Optional[Dict[str, Any]]:
    if not palette:
        return None
    closest = palette[0]
    min_dist = _color_distance(hex_val, closest["hex"])
    for color in palette:
        dist = _color_distance(hex_val, color["hex"])
        if dist < min_dist:
            min_dist = dist
            closest = color
    return {**closest, "distance": min_dist}


def _extract_hex_colors(text: Optional[str]) -> List[str]:
    if not text or not isinstance(text, str):
        return []
    return re.findall(r"#[0-9A-Fa-f]{6}", text)


def _extract_focal_length(text: Optional[str]) -> Optional[int]:
    if not text or not isinstance(text, str):
        return None
    m = re.search(r"(\d+)\s*mm", text, re.IGNORECASE)
    return int(m.group(1)) if m else None


def _extract_camera_angle(text: Optional[str]) -> Optional[int]:
    if not text or not isinstance(text, str):
        return None
    m = re.search(r"(\d+)", text)
    return int(m.group(1)) if m else None


# --------------------------------------------------------------------------
# Prompt validation (rule-based, no AI needed)
# --------------------------------------------------------------------------

def validate_prompt(
    prompt: Dict[str, Any],
    brand_style: Dict[str, Any],
) -> Dict[str, Any]:
    """
    Validate a product/image generation prompt against brand style rules.
    Returns {is_valid, compliance_score, violations, auto_fixes_available}.
    """
    violations: List[Dict[str, Any]] = []

    # --- Colour palette check ---
    color_scheme = prompt.get("color_scheme", "")
    if isinstance(color_scheme, dict):
        color_scheme = str(color_scheme)
    hex_colors = _extract_hex_colors(color_scheme)

    color_palette = brand_style.get("colorPalette", [])
    for hex_val in hex_colors:
        is_approved = any(
            _color_distance(c["hex"], hex_val) < 30 for c in color_palette
        )
        if not is_approved:
            closest = _find_closest_color(hex_val, color_palette)
            violations.append(
                {
                    "id": str(uuid.uuid4()),
                    "rule": "Color differs from brand palette (OK for trends)",
                    "category": "color",
                    "severity": "suggestion",
                    "detected": hex_val,
                    "allowed": ", ".join(
                        f"{c['name']} ({c['hex']})" for c in color_palette
                    )
                    or "No colors defined",
                    "message": f"Trend color {hex_val} differs from brand palette",
                    "autoFixAvailable": closest is not None,
                    "fixedValue": closest["hex"] if closest else None,
                }
            )

    # --- Camera settings ---
    camera = brand_style.get("cameraSettings", {})
    focal_length = _extract_focal_length(str(prompt.get("focal_length", "")))
    if focal_length is not None:
        fov = round(2 * math.atan(36 / (2 * focal_length)) * (180 / math.pi))
        fov_min = camera.get("fovMin", 20)
        fov_max = camera.get("fovMax", 80)
        if fov < fov_min or fov > fov_max:
            default_fov = camera.get("fovDefault", 50)
            suggested = round(
                36 / (2 * math.tan((default_fov * math.pi / 180) / 2))
            )
            violations.append(
                {
                    "id": str(uuid.uuid4()),
                    "rule": "Focal length must produce FOV within allowed range",
                    "category": "camera",
                    "severity": "warning",
                    "detected": f"{focal_length}mm (FOV ~{fov}°)",
                    "allowed": {"min": fov_min, "max": fov_max},
                    "message": f"Focal length {focal_length}mm produces FOV ~{fov}° outside range ({fov_min}°-{fov_max}°)",
                    "autoFixAvailable": True,
                    "fixedValue": f"{suggested}mm",
                }
            )

    angle = _extract_camera_angle(str(prompt.get("camera_angle", "")))
    if angle is not None:
        angle_min = camera.get("angleMin", 0)
        angle_max = camera.get("angleMax", 90)
        if angle < angle_min or angle > angle_max:
            violations.append(
                {
                    "id": str(uuid.uuid4()),
                    "rule": "Camera angle must be within allowed range",
                    "category": "camera",
                    "severity": "warning",
                    "detected": prompt.get("camera_angle"),
                    "allowed": {"min": angle_min, "max": angle_max},
                    "message": f"Camera angle outside range ({angle_min}°-{angle_max}°)",
                    "autoFixAvailable": True,
                    "fixedValue": f"{camera.get('angleDefault', 45)}° eye level",
                }
            )

    # --- Lighting ---
    lighting_cfg = brand_style.get("lightingConfig", {})
    lighting_raw = prompt.get("lighting", "")
    lighting_lower = (
        str(lighting_raw).lower()
        if isinstance(lighting_raw, str)
        else str(lighting_raw).lower()
    )
    color_temp = lighting_cfg.get("colorTemperature", 5000)
    if lighting_lower:
        if color_temp < 4500 and "cool" in lighting_lower:
            violations.append(
                {
                    "id": str(uuid.uuid4()),
                    "rule": "Lighting color temperature should match brand style",
                    "category": "lighting",
                    "severity": "suggestion",
                    "detected": "cool lighting",
                    "allowed": f"warm lighting ({color_temp}K)",
                    "message": f"Cool lighting detected but brand specifies warm ({color_temp}K)",
                    "autoFixAvailable": True,
                    "fixedValue": "warm, soft lighting with natural tones",
                }
            )
        elif color_temp > 5500 and "warm" in lighting_lower:
            violations.append(
                {
                    "id": str(uuid.uuid4()),
                    "rule": "Lighting color temperature should match brand style",
                    "category": "lighting",
                    "severity": "suggestion",
                    "detected": "warm lighting",
                    "allowed": f"cool lighting ({color_temp}K)",
                    "message": f"Warm lighting detected but brand specifies cool ({color_temp}K)",
                    "autoFixAvailable": True,
                    "fixedValue": "cool, neutral studio lighting",
                }
            )

    # --- Negative prompts ---
    negative_prompts = brand_style.get("negativePrompts", [])
    description_raw = prompt.get("description", "")
    description_lower = (
        str(description_raw).lower()
        if isinstance(description_raw, str)
        else str(description_raw).lower()
    )

    for neg in negative_prompts:
        if description_lower and neg.lower() in description_lower:
            violations.append(
                {
                    "id": str(uuid.uuid4()),
                    "rule": "Description must not contain forbidden terms",
                    "category": "prompt",
                    "severity": "critical",
                    "detected": neg,
                    "allowed": "Not allowed: " + ", ".join(negative_prompts),
                    "message": f'Description contains forbidden term: "{neg}"',
                    "autoFixAvailable": True,
                    "fixedValue": re.sub(neg, "", str(description_raw), flags=re.IGNORECASE).strip(),
                }
            )

    current_negatives = (prompt.get("negative_prompt") or "").lower()
    for neg in negative_prompts:
        if neg.lower() not in current_negatives:
            violations.append(
                {
                    "id": str(uuid.uuid4()),
                    "rule": "Negative prompt must include brand exclusions",
                    "category": "prompt",
                    "severity": "warning",
                    "detected": "Missing: " + neg,
                    "allowed": ", ".join(negative_prompts),
                    "message": f'Missing required negative prompt: "{neg}"',
                    "autoFixAvailable": True,
                    "fixedValue": neg,
                }
            )

    # --- Objects check ---
    for obj in prompt.get("objects", []):
        obj_desc = str(obj.get("description", "")).lower()
        for neg in negative_prompts:
            if neg.lower() in obj_desc:
                violations.append(
                    {
                        "id": str(uuid.uuid4()),
                        "rule": "Object description must not contain forbidden terms",
                        "category": "prompt",
                        "severity": "critical",
                        "detected": f'"{neg}" in object "{obj.get("name", "")}"',
                        "allowed": "Not allowed: " + ", ".join(negative_prompts),
                        "message": f'Object "{obj.get("name", "")}" contains forbidden term: "{neg}"',
                        "autoFixAvailable": True,
                        "fixedValue": re.sub(neg, "", str(obj.get("description", "")), flags=re.IGNORECASE).strip(),
                    }
                )

    # --- Scoring ---
    critical = sum(1 for v in violations if v["severity"] == "critical")
    warning = sum(1 for v in violations if v["severity"] == "warning")
    suggestion = sum(1 for v in violations if v["severity"] == "suggestion")

    score = max(0, min(100, 100 - critical * 25 - warning * 10 - suggestion * 3))

    return {
        "is_valid": critical == 0,
        "compliance_score": score,
        "violations": violations,
        "auto_fixes_available": sum(1 for v in violations if v.get("autoFixAvailable")),
    }


def get_compliance_badge(score: int) -> Dict[str, str]:
    if score >= 90:
        return {"label": "Excellent", "color": "text-emerald-400", "bgColor": "bg-emerald-500/20"}
    if score >= 75:
        return {"label": "Good", "color": "text-green-400", "bgColor": "bg-green-500/20"}
    if score >= 60:
        return {"label": "Fair", "color": "text-amber-400", "bgColor": "bg-amber-500/20"}
    if score >= 40:
        return {"label": "Poor", "color": "text-orange-400", "bgColor": "bg-orange-500/20"}
    return {"label": "Critical", "color": "text-red-400", "bgColor": "bg-red-500/20"}
